keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,418 @@
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine learning researcher I want to run a workload, comprising tens of
+pods which need to all run at the same time. Autoscaling the cluster and
+scheduling of all pods in this workload take a couple of minutes and I want to
+have a guarantee that other workloads scheduled within this time interval would
+not delay execution of my workload.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pods-scheduled`
+annotation to the jobs I run.
+
+#### Story 2
+
+As an administrator of a batch processing in my organization I want to be able
+to configure kueue quotas for on-demand resources a little higher than what is
+available in the quiet period. However, I don't want to risk two workloads
+deadlock in case of issues with provisioning nodes to satisfy the configured
+quota.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pods-scheduled`
+annotation to all jobs.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+The risk can be mitigated by using the Job's `ActiveDeadlineSeconds` timeout
+to terminate a started job, which results in finishing the associated workload.
+
+Still the mitigation might be problematic as the timeout needs to include both
+the scheduling and execution time for a job, thus needs to be higher than one
+would want.
+
+Thus, we consider introduction of timeout on reaching the `PodsScheduled`
+condition by a workload since its job start (see:
+[Timeout on reaching the PodsScheduled condition](#timeout-on-reaching-the-podsscheduled-condition))
+as a possible future improvement.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### PodsScheduled workload condition
+
+First, we introduce a new workload condition, called `PodsScheduled`, to indicate
+that the first batch of workload's pods is scheduled, more precisely:
+`job.status.ready + job.status.succeeded >= job.spec.parallelism`.
```

Comment:
Explain that job controllers are responsible for populating this condition and then discuss how we are doing it for v1.Job in a separate paragraph.

Reply:
The condition is  populated by the Kueue's Job reconciler. Rephrased and used the Job reconciler name vs job controller to avoid confusion.

Reply:
Thanks for the high-level perspective. 
Re 1: I renamed the uses of workload to job to avoid the confusion.
Re 2: Indeed, it could be an argument against setting the annotation on Job. Not sure if it outweights the other pros. I have updated the review: https://github.com/kubernetes-sigs/kueue/pull/433#discussion_r1033359206. Going to also put these into alternatives. Another possibility would be to keep the workload API field, but populate it based on the local queue configuration. WDYT? It would mean that the admin needs to set up a local queue with the configuration, so that the users sent their workloads there. On the other side, it means there is no need for the new Job annotation.

Reply:
Added the discussion into the notes section in the doc. 
After the note I'm leaning to using LocalQueue API fields (for opting in and timeout declaration).

Reply:
/hold let's hold until we have more opinions on this, in particular: Jobs annotation vs. LocalQueue API field.

Reply:
We have a long term plan to use LocalQueue as a place for workload defaulting, I wouldn't add it now though, lets give the feature time to soak a bit, so perhaps starting with an annotation on the Job API is a good first step.

Reply:
There are a couple of advantages which LocalQueue seems to have:
- it does not leak implementation to JobReconciler, allowing to keep it lean as a reference implementation of the CRD <> Workload integration (https://github.com/kubernetes-sigs/kueue/pull/433#discussion_r1034128187). 
- it would also be a natural place for the timeout, currently the opt-in and related timeout are on two opposite ends of the concepts. Then, it allows you to have two LocalQueues, so that the timeouts could be different for large and small jobs.

In terms of implementation complexity, I think it is similar, but we have already a webhook for validating LocalQueue, and if we want to validate the job annotation we would need to introduce a new one. 

I also think that in terms of the "first step" it also seems more natural to stay within Kueue concepts.

Reply:
A couple of thoughts/history:

1) The integration controllers should be lean, there is no disagreement, it is a principle we have been trying to maintain since day 1 :). 
2)  Any workload level API that we expose can be populated by the integration controller, we will not intentionally block that.
3) LocalQueue can be thought of as a place for defaulting workload parameters, this was one of the design goals of LocalQueue.

> it does not leak implementation to JobReconciler, allowing to keep it lean as a reference implementation of the CRD <> Workload integration (https://github.com/kubernetes-sigs/kueue/pull/433#discussion_r1034128187).

If we want this to continue to have this as a workload level API (with an extension in the future to be defaulted via LocalQueue), then this is not leaking. As for complexity, as I mentioned on the other comment, we will have that no matter what option we choose because we need to translate pod ready status to a condition on the workload. 

> I also think that in terms of the "first step" it also seems more natural to stay within Kueue concepts.

My thinking is that we want to be very careful when adding new fields to ClusterQueue and LocalQueue since those are user-facing apis and we need to be very careful when adding new fields to them. While the Workload API is intended as a gluing layer that we may have a bit more wiggle room to experiment with, while at the same time adding an annotation to the job api is simple to deprecate.